### PR TITLE
Login routes

### PR DESCRIPTION
**Pull request by moralesbang**
## Summary

- Added protected routes for no authenticated users.
- This PR also resolver [this card](https://trello.com/c/o0bOo78w).

## Screenshots

Aug-17-2019 20-51-12

## Trello Card

https://trello.com/c/kPxmDWf2

### CODE REVIEW

File: react/tic-tac-toe/src/app/components/Router/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { history } from '~redux/store'; // eslint-disable-line import/no-unresolved
+
+import CheckRoute from './components/CheckRoute';
+import { PUBLIC_ROUTES } from './constants'; // eslint-disable-line import/no-unresolved
+
+function Router() {
+  return (
+    <ConnectedRouter history={history}>
+      <Switch>
+        {PUBLIC_ROUTES.map((route) => <CheckRoute key={route.path} {...route} />)}
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/components/Router/index.js]: You should map all the routes, not just the public ones

File: react/tic-tac-toe/src/app/screens/Login/constants.js [hunk 2]
```diff
@@ -4,3 +4,14 @@ export const VALIDATIONS = {
   email: [FORM_VALIDATIONS.required, FORM_VALIDATIONS.email],
   password: [FORM_VALIDATIONS.required, FORM_VALIDATIONS.minLength(8)]
 };
+
+const ERRORS = {
+  unauthorized: 'Unauthorized'
+};
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/screens/Login/constants.js]: Is this object necessary?

File: react/tic-tac-toe/src/redux/auth/actions.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import AuthService from '~services/AuthService'; // eslint-disable-line import/no-unresolved
+
+export const actions = {
+  LOGIN_SUCCESS: '@@AUTH/LOGIN_SUCCESS',
+  LOGIN_FAILURE: '@@AUTH/LOGIN_FAILURE'
+};
+
+const actionCreators = {
+  login: (credentials) => async (dispatch) => {
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/redux/auth/actions.js]: I think you don't need the `()`

File: react/tic-tac-toe/src/services/AuthService.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import api from '~config/api'; // eslint-disable-line import/no-unresolved
+
+export default {
+  login: (credentials) => api.post('/login', credentials)
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/services/AuthService.js]: same

File: react/tic-tac-toe/src/redux/store.js [hunk 5]
```diff
@@ -1,5 +1,29 @@
...
-export default createStore(combineReducers({ form: formReducer }), composeEnhancers());
+export const history = createBrowserHistory();
+
+const rootReducer = combineReducers({
+  form: formReducer,
+  auth: authReducer,
+  router: connectRouter(history)
+});
+
+const configureStore = (preloadedState) => createStore(
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/redux/store.js]: same

File: react/tic-tac-toe/src/app/screens/Login/utils.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  if (!values.email) {
+    errors.email = 'Ingresa tu email';
+  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
+    errors.email = 'Selecciona una dirección de email válido';
+  }
+
+  if (!values.password) {
+    errors.password = 'Selecciona una contraseña';
+  } else if (values.password.length < 8) {
+    errors.password = 'Debe tener al menos 8 caracteres';
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/screens/Login/utils.js]: You are using the errors in spanish when the app is all in english

File: react/tic-tac-toe/src/app/screens/Login/utils.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+export const validate = values => {
+  const errors = {};
+  if (!values.email) {
+    errors.email = 'Ingresa tu email';
+  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
+    errors.email = 'Selecciona una dirección de email válido';
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/screens/Login/utils.js]: `Selecciona `? I think it should be `Ingrese`

File: react/tic-tac-toe/src/app/screens/Login/index.js [hunk 8]
```diff
@@ -1,44 +1,57 @@
...
 }
 
-Login.propTypes = {
-  handleSubmit: func.isRequired
-};
+const mapStateToProps = ({ auth }) => ({
+  loginError: auth.loginError
+});
+
+const mapDispatchToProps = (dispatch) => ({
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/screens/Login/index.js]: Remove unnecesary `()`

File: react/tic-tac-toe/src/app/screens/Login/index.js [hunk 9]
```diff
@@ -1,44 +1,57 @@
...
+        </button>
+      </form>
     </div>
   );
 }
 
-Login.propTypes = {
-  handleSubmit: func.isRequired
-};
+const mapStateToProps = ({ auth }) => ({
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/screens/Login/index.js]: Use `store` here. Normally the `mapStateToProps` tend to grow quickly and the destructuring isn't very scalable. Also, using store makes the path much more readable.
`store.auth.loginError`

File: react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      ? <RouteComponent />
+      : <Redirect to="/login" />;
+  }
+
+  render () {
+    return <Route path={this.props.path} render={this.renderRoute} />;
+  }
+}
+
+const mapDispatchToProps = (state) => ({
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js]: Remove uneccesary `()`

File: react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import React, { Component } from 'react';
+import { Redirect, Route } from 'react-router-dom';
+import { connect } from 'react-redux';
+import { string, func } from 'prop-types';
+
+class CheckRoute extends Component {
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js]: I think this should be a function Component

File: react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    const { component: RouteComponent, path, authToken } = this.props;
+
+    if (authToken) {
+      return path === '/login'
+        ? <Redirect to="/" />
+        : <RouteComponent />;
+    }
+
+    return path === '/login'
+      ? <RouteComponent />
```
Comment (LucasZibell) [inline react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js]: Is this correct? I'm not sure the logic is correct. If you have an `authToken` and the page ig login you are sending him to a component.

And the else is `if you don't have an authToken and you aren't going to the login then go to the login` 🤔

File: react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js (hunk 11 above)
Comment (moralesbang) [reply react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js]: I have refactored that component and now is a `Function Component`

File: react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js (hunk 12 above)
Comment (moralesbang) [reply react/tic-tac-toe/src/app/components/Router/components/CheckRoute/index.js]: That logic was update, and follow the idea exposed in the meeting with you.

File: react/tic-tac-toe/src/app/screens/Login/index.js (hunk 9 above)
Comment (moralesbang) [reply react/tic-tac-toe/src/app/screens/Login/index.js]: Nice! I gonna have this in count.

File: react/tic-tac-toe/src/app/components/Router/components/DefaultLayout/index.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import { Route, Switch } from 'react-router-dom';
+
+import { PROTECTED_ROUTES } from '../../constants';
+
+function DefaultLayout() {
```
[the remark this answers is not in the document]
Comment (moralesbang) [reply react/tic-tac-toe/src/app/components/Router/components/DefaultLayout/index.js]: I'm gonna think in a better name.

**Pull request closed without merging**